server/service/src/asset/insert.rs

```diff
@@ -0,0 +1,112 @@
+use super::{query::get_asset, validate::check_asset_exists};
+use crate::{service_provider::ServiceContext, SingleRecordError};
+use chrono::{NaiveDate, Utc};
+use repository::{
+    assets::{
+        asset::{AssetFilter, AssetRepository},
+        asset_row::{AssetRow, AssetRowRepository},
+    },
+    RepositoryError, StorageConnection, StringFilter,
+};
+
+#[derive(PartialEq, Debug)]
+pub enum InsertAssetError {
+    AssetAlreadyExists,
+    CreatedRecordNotFound,
+    DatabaseError(RepositoryError),
+    SerialNumberAlreadyExists,
+}
+
+pub struct InsertAsset {
+    pub id: String,
+    pub store_id: Option<String>,
+    pub name: String,
+    pub code: String,
+    pub serial_number: Option<String>,
+    pub catalogue_item_id: Option<String>,
+    pub installation_date: Option<NaiveDate>,
+    pub replacement_date: Option<NaiveDate>,
+}
+
+pub fn insert_asset(
+    ctx: &ServiceContext,
+    input: InsertAsset,
+) -> Result<AssetRow, InsertAssetError> {
+    let asset = ctx
+        .connection
+        .transaction_sync(|connection| {
+            validate(&input, connection)?;
+            let new_asset = generate(input);
+            AssetRowRepository::new(&connection).upsert_one(&new_asset)?;
+
+            get_asset(ctx, new_asset.id).map_err(InsertAssetError::from)
+        })
+        .map_err(|error| error.to_inner_error())?;
+    Ok(asset)
+}
+
+pub fn validate(
+    input: &InsertAsset,
+    connection: &StorageConnection,
+) -> Result<(), InsertAssetError> {
+    if check_asset_exists(&input.id, connection)?.is_some() {
+        return Err(InsertAssetError::AssetAlreadyExists);
+    }
+
+    // Check the serial number is unique (if present)
+    if let Some(serial_number) = &input.serial_number {
+        if AssetRepository::new(connection)
+            .query_one(AssetFilter::new().serial_number(StringFilter::equal_to(serial_number)))?
+            .is_some()
+        {
+            return Err(InsertAssetError::SerialNumberAlreadyExists);
+        }
+    }
+
+    // TODO: Do we require Name, and or Code? I'm assuming user can leave them as empty strings if they want?
```

Comment:
Question for the reviewer... Any other validations needed?

Reply:
Agree - the user can leave them as empty strings. It would be nice to have names in the list, I don't think we need to require them however